src/kernel.cu

```diff
@@ -0,0 +1,131 @@
+#include <cstdlib>
+#include <stdio.h>
+#include <float.h>
+
+#include "point.h"
+
+//Testing for CUDA errors
+inline cudaError_t checkCuda(cudaError_t result) {
+if (result != cudaSuccess) {
+fprintf(stderr, "CUDA Runtime Error: %s\n", cudaGetErrorString(result));
+}
+return result;
+}
+
+//Clustering Calculation
+__global__ 
+void my_Kernel_1(Point* points_d, int points_size, Point* c_d, int cluster_id) {
+    
+    //Get the index for the current point to work with
+    int Col = threadIdx.x + blockIdx.x * blockDim.x;
+    int Row = threadIdx.y + blockIdx.y * blockDim.y;
+    int index = Col + (Row * blockDim.x *gridDim.x); 
+    
+    if(index < points_size) {
+        Point p = points_d[index];
+        double dist = ((c_d->x - p.x) * (c_d->x - p.x) + (c_d->y - p.y) * (c_d->y - p.y) + (c_d->z - p.z) * (c_d->z - p.z));// point_distance(c, *p);
+        if (dist < p.min_dist) {
+            p.min_dist = dist;
+            p.cluster = cluster_id;
+        }
+        
+        points_d[index] = p;
+    }
+}
+
+__global__
+void my_Kernel_2(double max, Point* points_d, int points_size, int* n_points_d, double* sum_x_d, double* sum_y_d, double* sum_z_d) {
+    //Get the index of the point to work with
+    int Col = threadIdx.x + blockIdx.x * blockDim.x;
+    int Row = threadIdx.y + blockIdx.y * blockDim.y;
+    int index = Col + (Row * blockDim.x * gridDim.x); 
+    
+    if(index < points_size) {
+        Point p = points_d[index];
+        n_points_d[p.cluster] += 1;
+        sum_x_d[p.cluster] += p.x;
+        sum_y_d[p.cluster] += p.y;
+        sum_z_d[p.cluster] += p.z;
+
+        // reset distance
+        p.min_dist = max;
+        points_d[index] = p;
+    }
+}
+
+// Function that launches the CUDA kernel
+extern "C" void cuda_Kernel_1(Point* points_h, int points_size, Point c, int cluster_id) {
+    Point* points_d;
+    Point* c_d;
+    int new_points_size = sqrt(points_size); 
+
+
+    //Allocate device pointers and copy them to the device
+    checkCuda(cudaMalloc((void **) &points_d, sizeof(Point)*points_size));
```

Comment:
I don't know if this would make a big performance difference, but we should be able to call these cudaMalloc's only once at the beginning of the program and the cudaFree at the end. Then everytime we want to call the kernel function we would just do the cudaMemcpy and the kernel function.

Reply:
Now that I think of it, is there any reason why we couldn't keep the entire Points array in GPU memory the entire time? The constant copying back and forth is probably expensive